inc/functions-shortcodes.php

```diff
@@ -215,3 +218,50 @@ function members_login_form_shortcode() {
 
 	return wp_login_form( array( 'echo' => false ) );
 }
+
+/**
+ * Filters the login redirect URL to send failed logins back to the
+ * referrer with a query arg of `login=failed`.
+ *
+ * @since 3.2.18
+ *
+ * @param string $redirect_to    The redirect destination URL.
+ * @param string $request        The request URL.
+ * @param object $user           The user object.
+ * @return string                The redirect URL.
+ */
+function members_login_redirect( $redirect_to, $request, $user ) {
+    // If there's a valid referrer, and it's not the default log-in screen
+    if ( ! isset( $_POST['members_redirect_to'] ) ) {
+        return $redirect_to;
+    } elseif ( empty( $user ) || is_wp_error( $user ) ) {
+        wp_redirect( str_replace('?login=failed', '', $_SERVER['HTTP_REFERER']) . '?login=failed' );
+        die;
+    } else {
+        return str_replace('?login=failed', '', $_SERVER['HTTP_REFERER']);
+    }
+}
+
+/**
+ * Filters the login form bottom output to add an error message if the login has failed.
+ *
+ * @since 3.2.18
+ *
+ * @return string The HTML to output below the login form.
+ */
+function members_login_form_bottom() {
+    if (
+        isset( $_SERVER['HTTP_REFERER'] ) &&
```

Comment:
@ThemeGravity is any of this referrer checks necessary here? Can't we just always append these to the form?

Reply:
@cartpauj We use `login_form_bottom` filter hook from `wp_login_form` function to display hidden input so you are right that referrer checks are not needed. I removed them in the last commit.